File: android/app/src/main/java/de/repeatuntil/yata/domain/user/UserService.kt
package de.repeatuntil.yata.domain.user

import de.repeatuntil.yata.domain.common.DomainException
import de.repeatuntil.yata.domain.common.mapFailureToRepositoryException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class UserService(
    private val repository: UserRepository,
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
) {

    suspend fun getActiveUser(): Result<User?> {
        return withContext(ioDispatcher) {
            repository.findActiveUser()
                .mapFailureToRepositoryException()
        }
    }

    suspend fun getAllUsers(): Result<List<User>> {
        return withContext(ioDispatcher) {
            repository.getAllUsers()
                .mapFailureToRepositoryException()
        }
    }

    suspend fun addUser(user: User): Result<Unit> {
        return withContext(ioDispatcher) {
            if (user.isActive) {
                addActiveUser(user)
            } else {
                repository.addUser(user)
                    .mapFailureToRepositoryException()
            }
        }
    }

    private suspend fun addActiveUser(activeUser: User): Result<Unit> {
        assert(activeUser.isActive)
        return withContext(ioDispatcher) {
            deactivateOtherUsersForActiveUser(activeUser) { repository.addUser(it) }
        }
    }

    private suspend fun updateActiveUser(activeUser: User): Result<Unit> {
        assert(activeUser.isActive)
        return withContext(ioDispatcher) {
            deactivateOtherUsersForActiveUser(activeUser) { repository.updateUser(it) }
        }
    }

    private suspend fun deactivateOtherUsersForActiveUser(
        activeUser: User,
        actionForActiveUser: suspend (User) -> Result<Unit>
    ): Result<Unit> {
        assert(activeUser.isActive)
        return withContext(ioDispatcher) {
            repository.getAllUsers()
                .map { users ->
                    // Deactivate all other users
                    users.mapNotNull { if (it.id != activeUser.id) it.copy(isActive = false) else null }
                }
                // Update users
                .fold(
                    onSuccess = { repository.updateUsers(it) },
                    onFailure = { Result.failure(DomainException.RepositoryException(it)) }
                )
                // Run action for user
                .fold(
                    onSuccess = { actionForActiveUser(activeUser) },
                    onFailure = { Result.failure(DomainException.RepositoryException(it)) }
                )
                .mapFailureToRepositoryException()
        }
    }

    suspend fun updateUser(user: User): Result<Unit> {
        return withContext(ioDispatcher) {
            if (user.isActive) {
                updateActiveUser(user)
            } else {
                repository.updateUser(user)
                    .mapFailureToRepositoryException()
            }
        }
    }
}

File: ios/YATA/Domain/User/UserService.swift
//
//  UserService.swift
//  YATA
//
//  Created by Alpár Szotyori on 26.05.24.
//

import Foundation

class UserService {
    private let repository: UserRepository

    init(repository: UserRepository) {
        self.repository = repository
    }

    func getActiveUser() async -> Result<User?, UserServiceError> {
        return await repository.findActiveUser()
            .mapErrorToRepositoryError()
            .wrapDomainError()
    }

    func addUser(_ user: User) async -> Result<Void, UserServiceError> {
        if user.isActive {
            return await addActiveUser(user)
        } else {
            return await repository.addUser(user)
                .mapErrorToRepositoryError()
                .wrapDomainError()
        }
    }

    func getAllUsers() async -> Result<[User], UserServiceError> {
        return await repository.getAllUsers()
            .mapErrorToRepositoryError()
            .wrapDomainError()
    }

    func updateUser(_ user: User) async -> Result<Void, UserServiceError> {
        if user.isActive {
            return await updateActiveUser(user)
        } else {
            return await repository.updateUser(user)
                .mapErrorToRepositoryError()
                .wrapDomainError()
        }
    }
}

extension UserService {

    private func addActiveUser(_ activeUser: User) async -> Result<Void, UserServiceError> {
        assert(activeUser.isActive)
        return await deactivateOtherUsersForActiveUser(activeUser) { await repository.addUser($0) }
    }

    private func updateActiveUser(_ activeUser: User) async -> Result<Void, UserServiceError> {
        assert(activeUser.isActive)
        return await deactivateOtherUsersForActiveUser(activeUser) { await repository.updateUser($0) }
    }

    private func deactivateOtherUsersForActiveUser(_ activeUser: User,
                                                   actionForActiveUser: (User) async -> Result<Void, Error>) async
    -> Result<Void, UserServiceError> {
        assert(activeUser.isActive)

        let usersResult = await repository.getAllUsers()
            .map { users in
                users.filter { $0.id != activeUser.id }
                    .map { $0.copy(isActive: false) }
            }
        switch usersResult {
        case .success(let users):
            let updatedUsersResult = await repository.updateUsers(users)

            switch updatedUsersResult {
            case .success:
                return await actionForActiveUser(activeUser)
                    .mapErrorToRepositoryError()
                    .wrapDomainError()
            case .failure(let error):
                return Result.failure(error)
                    .mapErrorToRepositoryError()
                    .wrapDomainError()
            }
        case .failure(let error):
            return Result.failure(error)
                .mapErrorToRepositoryError()
                .wrapDomainError()
        }
    }
}

enum UserServiceError: Error {
    case domainError(_ error: DomainError)
}

fileprivate extension Result where Failure == DomainError {

    func wrapDomainError() -> Result<Success, UserServiceError> {
        return mapError { error in
            .domainError(error)
        }
    }
}
